apps/perl/5.18.2/man/man3/ops.3
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "ops 3"
.TH ops 3 "2013-11-04" "perl v5.18.2" "Perl Programmers Reference Guide"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
ops \- Perl pragma to restrict unsafe operations when compiling
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\&  perl \-Mops=:default ...    # only allow reasonably safe operations
\&
\&  perl \-M\-ops=system ...     # disable the \*(Aqsystem\*(Aq opcode
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Since the \f(CW\*(C`ops\*(C'\fR pragma currently has an irreversible global effect, it is
only of significant practical use with the \f(CW\*(C`\-M\*(C'\fR option on the command line.
.PP
See the Opcode module for information about opcodes, optags, opmasks
and important information about safety.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
Opcode, Safe, perlrun


lib/Agua/Ops.pm
use MooseX::Declare;

=head2

	PACKAGE		Agua::Ops
	
	PURPOSE
	
		CARRY OUT COMMON SYSTEM COMMANDS, INSTALL PACKAGES AND OTHER OPERATIONS

=cut

class Agua::Ops with (Agua::Ops::Ec2,
	Agua::Ops::Edit,
	Agua::Ops::Files,
	Agua::Ops::Git,
	Agua::Ops::Install,
	Agua::Ops::Nfs,
	Agua::Ops::Sge,
	Agua::Ops::Stager,
	Agua::Ops::Version,
	Agua::Common::App,
	Agua::Common::Aws,
	Agua::Common::Base,
	Agua::Common::Cluster,
	Agua::Common::Logger,
	Agua::Common::Package,
	Agua::Common::Parameter,
	Agua::Common::Source,
	Agua::Common::Ssh,
	Agua::Common::Util) {

use FindBin qw($Bin);
use lib "$Bin/..";

#### EXTERNAL MODULES
use Data::Dumper;
use File::Path;
use Getopt::Simple;
use Moose::Util qw(apply_all_roles);

#### INTERNAL MODULES
use Agua::OpsInfo;
use Conf::Yaml;
use Agua::Logger;

# Booleans
has 'warn'		=> ( isa => 'Bool', is 	=> 'rw', default	=>	1	);
has 'help'		=> ( isa => 'Bool', is  => 'rw', required	=>	0, documentation => "Print help message"	);
has 'backup'	=> ( isa => 'Bool', is  => 'rw', default	=>	0, documentation => "Automatically back up files before altering"	);

# Ints
has 'showlog'		=> ( isa => 'Int', is => 'rw', default 	=> 	2 	);  
has 'printlog'		=> ( isa => 'Int', is => 'rw', default 	=> 	2 	);
has 'sleep'			=> ( is  => 'rw', 'isa' => 'Int', default	=>	600	);
has 'upgradesleep'	=> ( is  => 'rw', 'isa' => 'Int', default	=>	600	);

# Strings
has 'opsrepo'	=> ( isa => 'Str|Undef', is => 'rw', required	=> 	0	);
has 'database'	=> ( isa => 'Str|Undef', is => 'rw', required	=> 	0	);
has 'user'
=> ( isa => 'Str|Undef', is => 'rw', required	=> 	0	);
has 'password'	=> ( isa => 'Str|Undef', is => 'rw', required	=> 	0	);
has 'conffile'	=> ( isa => 'Str|Undef', is => 'rw', required	=> 	0	);

has 'logfile'	=> ( isa => 'Str|Undef', is => 'rw', required	=>	0	);
has 'envfile'	=> ( isa => 'Str|Undef', is => 'rw', required	=>	0	);
has 'hostname'	=> ( isa => 'Str|Undef', is => 'rw', required	=>	0	);
has 'username'	=> ( isa => 'Str|Undef', is => 'rw', required	=>	0	);
#has 'sessionid'	=> ( isa => 'Str|Undef', is => 'rw', required	=>	0	);
has 'cwd'		=> ( isa => 'Str|Undef', is => 'rw', required	=>	0	);
has 'envars'	=> ( isa => 'Str|Undef', is => 'rw', required	=>	0	);
has 'tempdir'	=> ( isa => 'Str|Undef', is => 'rw', required	=>	0	);
has 'hubtype'	=> ( isa => 'Str|Undef', is => 'rw', default	=>	"github" );
has 'remote'	=> ( isa => 'Str|Undef', is => 'rw', required	=> 	0	);

#### Objects
has 'db'		=> ( isa => 'Agua::DBase::MySQL', is => 'rw', required => 0 );
has 'ssh'		=> ( isa => 'Agua::Ssh', is => 'rw', required	=>	0	);
has 'jsonparser'=> ( isa => 'JSON', is => 'rw', lazy => 1, builder => "setJsonParser" );

#### INITIALIZATION VARIABLES FOR Agua::Ops::GitHub
has 'owner'		=> ( isa => 'Str|Undef', is => 'rw', required	=>	0	);
has 'package'	=> ( isa => 'Str|Undef', is => 'rw', required	=>	0	);
has 'login'		=> ( isa => 'Str|Undef', is => 'rw', required	=>	0	);
has 'token'		=> ( isa => 'Str|Undef', is => 'rw', required	=>	0	);
has 'password'	=> ( isa => 'Str|Undef', is => 'rw', required	=>	0	);
has 'installdir'=> ( isa => 'Str|Undef', is => 'rw', required	=>	0	);
has 'version'	=> ( isa => 'Str|Undef', is => 'rw', required	=>	0	);
has 'keyfile'	=> ( isa => 'Str|Undef', is => 'rw', default	=>	''	);

has 'conf'	=> ( isa => 'Conf::Yaml', is => 'rw', lazy => 1, builder => "setConf" );
#has 'conf' 	=> (
#	is =>	'rw',
#	'isa' => 'Conf::Yaml',
#	default	=>	sub { Conf::Yaml->new();	}
#);

has 'opsinfo' 	=> (
	is 			=>	'rw',
	isa 		=> 'Agua::OpsInfo'
	#,
	#'builder'	=>	'setOpsInfo',
	#'lazy'		=>	1
);
has 'logger' 	=> (
	is 			=>	'rw',
	isa 		=> 'Agua::Logger',
	default	=>	sub { Agua::Logger->new({});	}
);

####/////}}}}

method BUILD ($hash) {
	#$self->logDebug("hash", $hash);
	$self->logDebug("");
	$self->initialise();
}

method initialise () {
	$self->logDebug("");
	
	#### OPEN LOGFILE IF DEFINED
	my $logfile = $self->logfile();
	$self->startLog($logfile) if defined $logfile;
	$self->logNote("logfile", $logfile)  if defined $logfile;

	$self->setEnv();

	$self->setSsh();
	
	$self->setHubType($self->hubtype());

	$self->setDbObject() if not defined $self->db() and $self->can('setDbObject') and defined $self->database() and not $self->database();
}

method setHubType ($hubtype) {
	return if not $hubtype;
	apply_all_roles($self, 'Agua::Ops::GitHub') if $hubtype eq "github";
}

method setSsh () {
	my $username	=	$self->username();
	my $hostname 	=	$self->hostname();
	my $keyfile 	= 	$self->keyfile();

	$self->logDebug("username", $username);
	$self->logDebug("hostname", $hostname);
	$self->logDebug("keyfile", $keyfile);
	
	return if not defined $self->username() or not $self->username();
	return if not defined $self->hostname() or not $self->hostname();
	return if not defined $self->keyfile() or not $self->keyfile();

	return $self->_setSsh($username, $hostname, $keyfile);
}

method setEnv {
	return if not defined $self->envfile();
	my $envfile = $self->envfile();
	my $lines = $self->fileLines($envfile);
	my $envars = '';
	foreach my $line ( @$lines )
	{
		next if not $line =~ /^\s*(\S+)\s+(\S+)\s*$/;
		my $key = $1;
		my $value = $2;
		$envars .= "$key=$value";
	}
	$self->envars($envars);
}

method args() {
	my $meta = $self->meta();

	my %option_type_map = (
		'Bool'     => '!',
		'Str'      => '=s',
		'Int'      => '=i',
		'Num'      => '=f',
		'ArrayRef' => '=s@',
		'HashRef'  => '=s%',
		'Maybe'    => ''
	);
	
	my $attributes = $self->fields();
	my $args = {};
	foreach my $attribute_name ( @$attributes )
	{
		my $attr = $meta->get_attribute($attribute_name);
		my $attribute_type  = $attr->{isa};
		$attribute_type =~ s/\|.+$//;
		$args -> {$attribute_name} = {  type => $option_type_map{$attribute_type}  };
	}

	return $args;
}

method localCommand ($command) {
#### RUN COMMAND LOCALLY
	$self->logDebug("command", $command);
	return `$command`;
}

method clearChangeDir () {
	$self->logNote();
	$self->cwd("");	
}
method changeDir ($directory) {
	$self->logNote("directory", $directory);
	my $cwd = $self->cwd();
	if ( defined $cwd and $directory !~ /^\// ) {
		$cwd =~ s/\/$//;
		$cwd = "$cwd/$directory";
		return 0 if not $self->foundDir($cwd);
		return 0 if not chdir($cwd);
		$self->cwd($cwd);
	}
	else {
		return 0 if not $self->foundDir($directory);
		return 0 if not chdir($directory);
		$self->cwd($directory);
		$cwd = $directory;
	}
	
	return 1;
}

method runCommand ($command) {
#### RUN COMMAND LOCALLY OR ON REMOTE HOST
	#### ADD ENVIRONMENT VARIABLES IF EXIST
	print "Agua::Ops::runCommand: $command\n";
	$self->showlog(4);
	
	$command = $self->envars() . $command if defined $self->envars();
	$command = "cd ".  $self->cwd() . "; " . $command if defined $self->cwd() and $self->cwd();
	$self->logDebug("command", $command);

	if ( defined $self->ssh() ) {
		$self->logDebug("DOING ssh->execute($command)");
		return $self->ssh()->execute($command);
	}
	else {
		my $stdoutfile = "/tmp/$$.out";
		my $stderrfile = "/tmp/$$.err";
		my $output = '';
		my $error = '';
		
		#### TAKE REDIRECTS IN THE COMMAND INTO CONSIDERATION
		if ( $command =~ />\s+/ ) {
			#### DO NOTHING, ERROR AND OUTPUT ALREADY REDIRECTED
			if ( $command =~ /\s+&>\s+/
				or ( $command =~ /\s+1>\s+/ and $command =~ /\s+2>\s+/)
				or ( $command =~ /\s+1>\s+/ and $command =~ /\s+2>&1\s+/) ) {
				print `$command`;
			}
			#### STDOUT ALREADY REDIRECTED - REDIRECT STDERR ONLY
			elsif ( $command =~ /\s+1>\s+/ or $command =~ /\s+>\s+/ ) {
				$command .= " 2> $stderrfile";
				print `$command`;
				$error = `cat $stderrfile`;
			}
			#### STDERR ALREADY REDIRECTED - REDIRECT STDOUT ONLY
			elsif ( $command =~ /\s+2>\s+/ or $command =~ /\s+2>&1\s+/ ) {
				$command .= " 1> $stdoutfile";
				print `$command`;
				$output = `cat $stdoutfile`;
			}
		}
		else {
			$command .= " 1> $stdoutfile 2> $stderrfile";
			print `$command`;
			$output = `cat $stdoutfile`;
			$error = `cat $stderrfile`;
		}
		
		$self->logNote("output", $output) if $output;
		$self->logNote("error", $error) if $error;
		
		##### CHECK FOR PROCESS ERRORS
		$self->logError("Error with command: $command ... $@") and exit if defined $@ and $@ ne "" and $self->can('warn') and not $self->warn();

		#### CLEAN UP
		`rm -fr $stdoutfile`;
		`rm -fr $stderrfile`;
		chomp($output);
		chomp($error);
		
		return $output, $error;
	}
}

method chompCommand ($command) {
#### RETURN CHOMPED RESULT OF runCommand
	my ($output, $error) = $self->runCommand($command);
	return if not defined $output;
	$self->logNote("output", $output);
	$self->logNote("error", $error);
	$output =~ s/\s+$//;
	$error =~ s/\s+$//;
	$self->logNote("Returning '$output', '$error'");
	return ($output, $error);
}


method repeatCommand ($command, $sleep, $tries, $errorregex) {
#### REPEATEDLY TRY SYSTEM CALL UNTIL A NON-ERROR RESPONSE IS RECEIVED
	$errorregex = "(error|Error|ERROR)" if not defined $errorregex; 
	$self->logDebug("command", $command);
	$self->logDebug("sleep", $sleep);
	$self->logDebug("tries", $tries);
	$self->logDebug("errorregex", $errorregex);
	
	my $output;
	my $error;
	my $error_message = 1;
	while ( $error_message and $tries )
	{
		($output, $error) = $self->runCommand($command);
		$self->logDebug("output", $output);
		$self->logDebug("error", $error);
		
		use re 'eval';	# EVALUATE AS REGEX
		$error_message = $output =~ /$errorregex/;
		$self->logDebug("error_message", $error_message) if not $output;
		no re 'eval';# STOP EVALUATING AS REGEX

		#### DECREMENT TRIES AND SLEEP
		$tries--;
		$self->logDebug("$tries tries left.")  if not $output and not $error;
		$self->logDebug("sleeping $sleep seconds") and sleep($sleep) if $error_message;
	}
	$self->logDebug("Returning output", $output);
	$self->logDebug("Returning error", $error);
	
	return $output, $error;
}

#### CONTROL METHODS
method timeoutCommand ($command, $timeout) {
	eval {
		local %SIG;
		$SIG{ALRM}=	sub{ print "timeout reached, after $timeout seconds!\n"; die; };
		alarm $timeout;
		return $self->runCommand($command);
		
		alarm 0;
	};

	$self->logDebug("end of timeout returning 0");
	alarm 0;
	
}

method repeatTry ($command, $regex, $tries, $sleep) {
#### REPEAT COMMAND UNTIL NON-ERROR RETURNED OR LIMIT REACHED
	$self->logDebug("Agua::Ops::repeatTry(command)");
	$self->logDebug("command", $command);
	$self->logDebug("sleep", $sleep);
	$self->logDebug("tries", $tries);
	
	my $result = '';	
	my $error = 1;
	while ( $error and $tries )
	{
		open(COMMAND, $command) or die "Can't exec command: $command\n";
		while(<COMMAND>) {
			$result .= $_;
		}
		close (COMMAND);
		$self->logDebug("result", $result);

		use re 'eval';	# EVALUATE AS REGEX
		$error = $result =~ /$regex/;
		$self->logDebug("error", $error);
		no re 'eval';# STOP EVALUATING AS REGEX

		#### DECREMENT TRIES AND SLEEP
		$tries--;
		$self->logDebug("$tries tries left. Sleeping $sleep seconds") if $error;
		$self->logDebug("current datetime: ");
		$self->logDebug(`date`);

		sleep($sleep) if $error;
	}
	$self->logDebug("Returning result", $result);
	
	return $result;
}

####s USER INPUT METHODS
method yes ($message, $max_times) {    
    $/ = "\n";
    my $input = <STDIN>;
    my $counter = 0;
    while ( $input !~ /^Y$/i and $input !~ /^N$/i )
    {
    	if ( $counter > $max_times )	{
			$self->logCritical("Exceeded 10 tries. Exiting.");
		}
    	
    	$self->logCritical("$message");
    	$input = <STDIN>;
    	$counter++;
    }	

    if ( $input =~ /^N$/i )	{	return 0;	}
    else {	return 1;	}
}

#### DATABASE METHODS
method setDbObject () {
	my $database 	= $self->database() || $self->conf()->getKey("database", "DATABASE");
	my $user		= $self->user() || $self->conf()->getKey("database", "USER");
	my $password	= $self->password() || $self->conf()->getKey("database", "PASSWORD");
	$self->logDebug("database", $database);
	$self->logDebug("user", $user);
	#$self->logDebug("password", $password);

   #### CREATE DB OBJECT USING DBASE FACTORY
    my $db = Agua::DBaseFactory->new( 'MySQL',
        {
			database	=>	$database,
            user      	=>  $user,
            password  	=>  $password,
			logfile		=>	$self->logfile(),
			showlog		=>	2,
			printlog	=>	2
        }
    ) or die "Can't create database object to create database: $database. $!\n";

	$self->db($db);
}





}

